test/integration/hds_integration_test.cc

```diff
@@ -76,6 +87,29 @@ class HdsIntegrationTest : public Grpc::VersionedGrpcClientIntegrationParamTest,
     RELEASE_ASSERT(result, result.message());
   }
 
+  Network::TransportSocketFactoryPtr createUpstreamTlsContext() {
+    envoy::extensions::transport_sockets::tls::v3::DownstreamTlsContext tls_context;
+    const std::string yaml = absl::StrFormat(
+        R"EOF(
+common_tls_context:
+  tls_certificates:
+  - certificate_chain: { filename: "%s" }
+    private_key: { filename: "%s" }
+  validation_context:
+    trusted_ca: { filename: "%s" }
+require_client_certificate: true
+)EOF",
+        TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcert.pem"),
+        TestEnvironment::runfilesPath("test/config/integration/certs/upstreamkey.pem"),
+        TestEnvironment::runfilesPath("test/config/integration/certs/cacert.pem"));
+    TestUtility::loadFromYaml(yaml, tls_context);
+    auto cfg = std::make_unique<Extensions::TransportSockets::Tls::ServerContextConfigImpl>(
+        tls_context, factory_context_);
+    static Stats::Scope* upstream_stats_store = new Stats::IsolatedStoreImpl();
+    return std::make_unique<Extensions::TransportSockets::Tls::ServerSslSocketFactory>(
+        std::move(cfg), context_manager_, *upstream_stats_store, std::vector<std::string>{});
```

Comment:
Possibly a refactor candidate with the same method in `transport_socket_match_integration_test.cc`.

Reply:
Hi Harvey, I went ahead and did a small refactor, moving this function up into HttpIntegration test, because I noticed a lot of tests that inherit from this class use this. You mentioned specifically `transport_socket_match_integration_test` so I reflected this change in this test as well. However, in my search I also found the following tests or files use the same or similar function:

`alpn_selection_integration_test.cc`
`sds_dynamic_integration_test.cc` and `sds_static_integration_test.cc`
`xfcc_integration_test.cc`

I think fixing all of these is a lot of extra code in this PR, would you like me to submit another PR that refactors this?

Reply:
Separate PR would be a bit cleaner, but up to you, the present PR isn't too big right now.

Reply:
@htuch I think I'll do a separate PR, after even more investigation it seems like there is a bit more cleanup needed here than I originally thought, because there are a few more places where this code exists. There is an available implementation in ssl_utility.cc it seems, and a few more tests use their own implementation. I want to create a PR that localizes all of these calls to the ssl_utility.cc version, and the HttpIntegrationTest version I just added directing its call to the ssl_utility version using its member variables.